File: OtherLanguages/SnakeGame.nim
# snake_game.nim
# Author: Sanne Karibo
# Simple console Snake Game in Nim

import os, terminal, strutils, times, random

const
  width = 64
  height = 24
  maxLength = 100

type
  Point = object
    x, y: int

var
  snake: array[0..maxLength-1, Point]
  length = 5
  dir = 1  # 1=right,2=down,3=left,4=up
  food: Point
  score = 0
  gameOver = false

proc clearScreen() =
  stdout.write("\x1b[2J\x1b[H")

proc generateFood() =
  while true:
    food.x = rand(width)
    food.y = rand(height)
    var collide = false
    for i in 0..<length:
      if snake[i] == food:
        collide = true
        break
    if not collide: break

proc draw() =
  clearScreen()
  echo "Score: ", score
  for y in 0..<height:
    for x in 0..<width:
      var printed = false
      for i in 0..<length:
        if snake[i].x == x and snake[i].y == y:
          if i == 0:
            stdout.write("O")
          else:
            stdout.write("o")
          printed = true
          break
      if not printed:
        if food.x == x and food.y == y:
          stdout.write("X")
        else:
          stdout.write(" ")
    stdout.write("\n")

proc moveSnake() =
  if gameOver: return
  var newHead = snake[0]
  case dir:
  of 1: newHead.x += 1
  of 2: newHead.y += 1
  of 3: newHead.x -= 1
  of 4: newHead.y -= 1

  if newHead.x < 0 or newHead.x >= width or newHead.y < 0 or newHead.y >= height:
    gameOver = true
    return

  for i in 0..<length:
    if snake[i] == newHead:
      gameOver = true
      return

  # Shift snake body
  for i in countdown(length-1, 1):
    snake[i] = snake[i-1]

  snake[0] = newHead

  if newHead == food:
    if length < maxLength:
      length.inc()
      snake[length-1] = snake[length-2]
    score += 10
    generateFood()

proc readKey(): char =
  let oldEcho = terminal.getEcho()
  let oldRaw = terminal.getRaw()
  terminal.setEcho(false)
  terminal.setRaw(true)
  var c = stdin.readChar()
  terminal.setEcho(oldEcho)
  terminal.setRaw(oldRaw)
  return c

proc main() =
  # Initialize snake
  for i in 0..<length:
    snake[i] = Point(x: 20 - i, y: 10)
  generateFood()
  clearScreen()
  echo "Snake Game - Created by Sanne Karibo"
  echo "Use WASD keys to move. Press 'q' to quit. Press any key to start..."
  discard readKey()

  while not gameOver:
    draw()
    if stdin.ready():
      let c = readKey()
      case c:
      of 'd': if dir != 3: dir = 1
      of 's': if dir != 4: dir = 2
      of 'a': if dir != 1: dir = 3
      of 'w': if dir != 2: dir = 4
      of 'q': gameOver = true
      else: discard
    moveSnake()
    sleep(100)

  draw()
  echo "Game Over! Final Score: ", score

when isMainModule:
  main()

File: OtherLanguages/SnakeGame.swift
//
// SnakeGame.swift
// Author: Sanne Karibo
// Simple console Snake game in Swift
//

import Foundation

let width = 64
let height = 24
let maxLength = 100

enum Direction: Int {
    case right = 1, down, left, up
}

struct Point: Equatable {
    var x: Int
    var y: Int
}

class SnakeGame {
    var snake: [Point] = []
    var direction: Direction = .right
    var food: Point
    var score = 0
    var gameOver = false

    init() {
        // Initialize snake in the middle
        for i in 0..<5 {
            snake.append(Point(x: 20 - i, y: 10))
        }
        food = Point(x: Int.random(in: 0..<width), y: Int.random(in: 0..<height))
    }

    func generateFood() {
        food = Point(x: Int.random(in: 0..<width), y: Int.random(in: 0..<height))
        while snake.contains(food) {
            food = Point(x: Int.random(in: 0..<width), y: Int.random(in: 0..<height))
        }
    }

    func moveSnake() {
        guard !gameOver else { return }
        var newHead = snake[0]
        switch direction {
        case .right: newHead.x += 1
        case .down: newHead.y += 1
        case .left: newHead.x -= 1
        case .up: newHead.y -= 1
        }

        // Check collisions
        if newHead.x < 0 || newHead.x >= width || newHead.y < 0 || newHead.y >= height {
            gameOver = true
            return
        }
        if snake.contains(newHead) {
            gameOver = true
            return
        }

        snake.insert(newHead, at: 0)

        if newHead == food {
            score += 10
            generateFood()
        } else {
            snake.removeLast()
        }
    }

    func draw() {
        var grid = Array(repeating: Array(repeating: " ", count: width), count: height)

        for p in snake {
            if p == snake[0] {
                grid[p.y][p.x] = "O" // Head
            } else {
                grid[p.y][p.x] = "o" // Body
            }
        }
        grid[food.y][food.x] = "X"

        print("\u{001B}[2J") // Clear screen

        print("Score: \(score)")
        for row in grid {
            print(row.joined())
        }
        if gameOver {
            print("\nGame Over! Final Score: \(score)")
            print("Press Enter to exit...")
        }
    }

    func start() {
        print("Snake Game - Created by Sanne Karibo")
        print("Use WASD keys to move. Press Enter to start...")
        _ = readLine()

        while !gameOver {
            draw()

            // Non-blocking input hack: just blocking for simplicity
            if let input = readLine() {
                switch input.lowercased() {
                case "d": if direction != .left { direction = .right }
                case "s": if direction != .up { direction = .down }
                case "a": if direction != .right { direction = .left }
                case "w": if direction != .down { direction = .up }
                default: break
                }
            }

            moveSnake()

            usleep(150_000) // 150ms delay
        }
        draw()
        _ = readLine()
    }
}

let game = SnakeGame()
game.start()
